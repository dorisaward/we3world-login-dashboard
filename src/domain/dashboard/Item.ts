export type Item = {
  /**
   * Unique identifier for the Item
   */
  id: string;

  /**
   * Human readable name of the Item
   */
  name: string;

  /**
   * Human readable description of the Item
   */
  description: string;
};
