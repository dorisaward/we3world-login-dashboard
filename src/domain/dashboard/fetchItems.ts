/**
 * @returns the {@link items} to be displayed
 */
export const fetchItems = async () => {
  try {
    const response = await fetch('/api/items');
    const responseJson = await response.json();
    return responseJson;
  } catch (error) {
    console.error(error);
  }
};
