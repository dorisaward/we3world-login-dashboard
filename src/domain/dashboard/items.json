[
  {
    "id": "0",
    "name": "Item 1",
    "description": "Description aesrdthfjfdhgzxchjvn srdthsrtrdhf srdthfyg"
  },
  {
    "id": "1",
    "name": "Item 2",
    "description": "chgncgnbdhthdt sxchncgbdxgcnb xfrhfyjmfyhm xfgtsrtgsdrg"
  },
  {
    "id": "2",
    "name": "Item 3",
    "description": "zxfhbvcbngnhjfdtj aesrdthfjfdhgzxchjvn srdthsrtrdhf srdthfyg"
  }
]