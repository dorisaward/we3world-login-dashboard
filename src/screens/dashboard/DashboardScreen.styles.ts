import {StyleSheet} from 'react-native';

export const styles = StyleSheet.create({
  container: {
    height: '100%',
    padding: 10,
    flex: 1,
  },
  activityIndicaton: {
    flex: 1,
  },
});
