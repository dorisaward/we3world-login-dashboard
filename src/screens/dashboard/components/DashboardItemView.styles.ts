import {StyleSheet} from 'react-native';

export const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 10,
    borderColor: 'black',
    borderWidth: 1,
    borderRadius: 10,
    backgroundColor: 'lavenderblush',
    marginBottom: 10,
  },
  name: {
    fontSize: 18,
    color: 'mediumvioletred',
  },
  description: {
    fontSize: 15,
    color: 'deeppink',
  },
});
