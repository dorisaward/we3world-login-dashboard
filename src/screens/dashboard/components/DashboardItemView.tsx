import React from 'react';
import {ListRenderItem, Text, View} from 'react-native';
import {styles} from './DashboardItemView.styles';
import {Item} from '../../../domain/dashboard/Item';

export const DashboardItemView: ListRenderItem<Item> = ({
  item: {name, description},
}) => (
  <View style={styles.container}>
    <Text style={styles.name}>{name}</Text>
    <Text style={styles.description}>{description}</Text>
  </View>
);
