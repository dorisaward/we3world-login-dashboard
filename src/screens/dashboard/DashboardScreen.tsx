import React, {useCallback, useEffect} from 'react';
import {ActivityIndicator, Button, FlatList, View} from 'react-native';
import {styles} from './DashboardScreen.styles';
import {Item} from '../../domain/dashboard/Item';
import {DashboardItemView} from './components/DashboardItemView';
import {useAppDispatch, useAppSelector} from '../../domain/redux/hooks';
import {fetchItems} from '../../domain/dashboard/fetchItems';
import {itemsSelector} from '../../domain/redux/reducers/itemsReducer';
import {userActions} from '../../domain/redux/reducers/userReducer';

export const DashboardScreen = (): React.JSX.Element => {
  const dispatch = useAppDispatch();
  const {items, loading} = useAppSelector(itemsSelector);

  useEffect(() => {
    dispatch(fetchItems() as any);
  }, [dispatch]);

  const handleLogout = useCallback(
    () => dispatch(userActions.logout()),
    [dispatch],
  );

  return (
    <View style={styles.container}>
      {loading ? (
        <ActivityIndicator
          testID="dashboard-screen-spinner"
          color={'mediumvioletred'}
          style={styles.activityIndicaton}
          size={'large'}
        />
      ) : (
        <FlatList
          data={items}
          renderItem={DashboardItemView}
          keyExtractor={keyExtractor}
        />
      )}
      <Button title="logout" onPress={handleLogout} />
    </View>
  );
};

const keyExtractor = (item: Item) => item.id;
