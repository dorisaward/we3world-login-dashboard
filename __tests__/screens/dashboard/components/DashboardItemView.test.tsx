import React from 'react';
import {render} from '@testing-library/react-native';
import {DashboardItemView} from '../../../../src/screens/dashboard/components/DashboardItemView';
import {ListRenderItemInfo} from 'react-native';
import {Item} from '../../../../src/domain/dashboard/Item';

it('should render', () => {
  // Given
  const item: Item = {
    id: 'test-id',
    name: 'test-name',
    description: 'test-description',
  };
  const renderable = <DashboardItemView item={item} {...otherProps} />;

  // When
  const {toJSON} = render(renderable);

  // Then
  expect(toJSON()).toMatchSnapshot();
});

const otherProps: Pick<ListRenderItemInfo<Item>, 'separators' | 'index'> = {
  separators: {
    highlight: jest.fn(),
    unhighlight: jest.fn(),
    updateProps: jest.fn(),
  },
  index: 0,
};
