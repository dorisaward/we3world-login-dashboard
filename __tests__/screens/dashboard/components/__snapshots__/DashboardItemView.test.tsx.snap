// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<View
  style={
    {
      "backgroundColor": "lavenderblush",
      "borderColor": "black",
      "borderRadius": 10,
      "borderWidth": 1,
      "flex": 1,
      "marginBottom": 10,
      "padding": 10,
    }
  }
>
  <Text
    style={
      {
        "color": "mediumvioletred",
        "fontSize": 18,
      }
    }
  >
    test-name
  </Text>
  <Text
    style={
      {
        "color": "deeppink",
        "fontSize": 15,
      }
    }
  >
    test-description
  </Text>
</View>
`;
